src/gtest/test_checktransaction.cpp

```diff
@@ -510,5 +510,296 @@ TEST(checktransaction_tests, ScCertVersion) {
+        <<" nonEmptyOriginalTx.GetHash() "<<nonEmptyOriginalTx.GetHash().ToString()
+        <<" revertedTxByAssignOp.GetHash() "<<revertedTxByAssignOp.GetHash().ToString();
+}
+
+TEST(TransactionManipulation, ExtendingTransactionOuts) {
+    CMutableTransaction mutTx;
+    EXPECT_TRUE(mutTx.getVout().size() == 0);
+
+    unsigned int OutNum = 10;
+    for(unsigned int idx = 0; idx < OutNum; ++idx)
+        mutTx.addOut(CTxOut(CAmount(idx),CScript()));
+
+    CTransaction txOutOnly = mutTx;
+    EXPECT_TRUE(txOutOnly.GetVout().size() == OutNum);
+    for(unsigned int idx = 0; idx < OutNum; ++idx)
+        EXPECT_FALSE(txOutOnly.IsBackwardTransfer(idx))<<"Output at pos "<<idx<<"wrongly marked as bwt";
+
+    unsigned int BwtNum = 7;
+    for(unsigned int idx = 0; idx < BwtNum; ++idx)
+        mutTx.addBwt(CTxOut(CAmount(idx + OutNum),CScript()));
+
+    CTransaction txBwtAttempt = mutTx;
+    EXPECT_TRUE(txBwtAttempt.GetVout().size() == OutNum);
+    for(unsigned int idx = 0; idx < OutNum; ++idx)
+        EXPECT_FALSE(txBwtAttempt.IsBackwardTransfer(idx))<<"Output at pos "<<idx<<"wrongly marked as bwt";
+}
+
+TEST(CertificateManipulation, EmptyCertTransformationToMutableIsNotReversible) {
+    // CopyCtor -> CopyCtor
+    CScCertificate        EmptyOriginalCert;
+    CMutableScCertificate mutByCopyCtor(EmptyOriginalCert);
+    CScCertificate        revertedCertByCopyCtor(mutByCopyCtor);
+
+    EXPECT_FALSE(EmptyOriginalCert == revertedCertByCopyCtor);
+    EXPECT_TRUE(EmptyOriginalCert.GetHash().IsNull());
+    EXPECT_FALSE(revertedCertByCopyCtor.GetHash().IsNull());
+    EXPECT_TRUE(EmptyOriginalCert.nFirstBwtPos == 0);
+    EXPECT_TRUE(revertedCertByCopyCtor.nFirstBwtPos == 0);
+
+    // AssignOp -> CopyCtor
+    CMutableScCertificate mutByAssignOp;
+    mutByAssignOp = EmptyOriginalCert;
+    CScCertificate revertedTxFromAssignement(mutByCopyCtor);
+
+    EXPECT_FALSE(EmptyOriginalCert == revertedTxFromAssignement);
+    EXPECT_TRUE(EmptyOriginalCert.GetHash().IsNull());
+    EXPECT_FALSE(revertedTxFromAssignement.GetHash().IsNull());
+    EXPECT_TRUE(revertedTxFromAssignement.nFirstBwtPos == 0);
+
+    // CopyCtor -> AssignOp
+    CScCertificate revertedTxByAssignOp;
+    revertedTxByAssignOp = mutByCopyCtor;
+
+    EXPECT_FALSE(EmptyOriginalCert == revertedTxByAssignOp);
+    EXPECT_TRUE(EmptyOriginalCert.GetHash().IsNull());
+    EXPECT_FALSE(revertedTxByAssignOp.GetHash().IsNull());
+    EXPECT_TRUE(revertedTxByAssignOp.nFirstBwtPos == 0);
+}
+
+TEST(CertificateManipulation, NonEmptyCertTransformationToMutableIsReversible) {
+    //create non-empty transaction
+    CMutableScCertificate helperMutCert;
+    unsigned int OutNum = 10;
+    for(unsigned int idx = 0; idx < OutNum; ++idx)
+        helperMutCert.addOut(CTxOut(CAmount(idx),CScript()));
+
+    unsigned int bwtOut = 3;
+    CScript bwtScript = CScript() << OP_DUP << OP_HASH160 << ToByteVector(uint160()) << OP_EQUALVERIFY << OP_CHECKSIG;
```

Comment:
@alsala  I am currently forced to create a dummy bwtScript since, in subsequent transformation from mutable cert to cert, GetHash() is called and a transformation to CBackwardTransfer is carried out; this fails if script is empty. Should instead map an empty CTxOut to an empty CBackwardTransfer?

Reply:
Yes, I agree